filters/include/pcl/filters/experimental/cartesian_filter.h

```diff
@@ -0,0 +1,144 @@
+   * voxel to be used
+   */
+  inline void
+  setMinimumPointsNumberPerVoxel(const std::size_t min_points_per_voxel)
+  {
+    min_points_per_voxel_ = min_points_per_voxel;
+  }
+
+  /** \brief Return the minimum number of points required for a voxel to be used.
+   */
+  inline std::size_t
+  getMinimumPointsNumberPerVoxel() const
+  {
+    return min_points_per_voxel_;
+  }
+
+  /** \brief Hash a point to the index of a 3D grid cell.
+   * \param[in] pt the point to hash
+   * \param[in] inverse_leaf_size the array of 1/leaf_size
+   * \param[in] min_b the minimum bin coordinate
+   * \param[in] divb_mul_y division multiplier of the second dimension
+   * \param[in] divb_mul_z division multiplier of the third dimension
+   */
+  inline std::size_t
+  hashPoint(const PointT& pt,
+            const Eigen::Array4f& inverse_leaf_size,
+            const Eigen::Vector4i& min_b,
+            const std::size_t divb_mul_y,
+            const std::size_t divb_mul_z) const
+  {
+    const std::size_t ijk0 = std::floor(pt.x * inverse_leaf_size[0]) - min_b[0];
+    const std::size_t ijk1 = std::floor(pt.y * inverse_leaf_size[1]) - min_b[1];
+    const std::size_t ijk2 = std::floor(pt.z * inverse_leaf_size[2]) - min_b[2];
+    return ijk0 + ijk1 * divb_mul_y + ijk2 * divb_mul_z;
+  }
+
+  /** \brief Hash a point to the index of a 2D grid cell.
+   * \param[in] pt the point to hash
+   * \param[in] inverse_leaf_size the array of 1/leaf_size
+   * \param[in] min_b the minimum bin coordinate
+   * \param[in] divb_mul_y division multiplier of the second dimension
+   */
+  inline std::size_t
+  hashPoint(const PointT& pt,
+            const Eigen::Array4f& inverse_leaf_size,
+            const Eigen::Vector4i& min_b,
+            const std::size_t divb_mul_y) const
+  {
+    const std::size_t ijk0 = std::floor(pt.x * inverse_leaf_size[0]) - min_b[0];
+    const std::size_t ijk1 = std::floor(pt.y * inverse_leaf_size[1]) - min_b[1];
+    return ijk0 + ijk1 * divb_mul_y;
+  }
+
+  /** \brief Check if the leaf size is too small, given the range of the point cloud
+   * \param[in] min_b the minimum bin coordinate
+   * \param[in] min_b the minimum bin coordinate
+   * \param[in] inverse_leaf_size the array of 1/leaf_size
+   * \note Return false if hash range is larger than the range of size_t as the unsigned
+   * int "wrap around" effect will occur during hashing a point
+   */
+  std::size_t
+  checkHashRange(const Eigen::Vector4f& min_p,
+                 const Eigen::Vector4f& max_p,
+                 const Eigen::Array4f& inverse_leaf_size) const
+  {
+    // Pass inverse_leaf_size[2] = 0 for checking 2D grids
+    const std::size_t dx = std::floor((max_p[0] - min_p[0]) * inverse_leaf_size[0]) + 1;
+    const std::size_t dy = std::floor((max_p[1] - min_p[1]) * inverse_leaf_size[1]) + 1;
+    const std::size_t dz = std::floor((max_p[2] - min_p[2]) * inverse_leaf_size[2]) + 1;
```

Comment:
Should we check for overflow here as well? The actual multiplication that happens is not `(max_p - min_p) * inverse_leaf_size` but is `map_p.array().abs().max(min_p.array().abs()) * inverse_leaf_size`

We wouldn't catch that overflow with this logic

Reply:
What about casting everything to `double` before multiplications, and then check in `double` domain?

```cpp
const double max = static_cast<double>(std::numeric_limits<std::size_t>::max());

const double dx =
    std::floor(static_cast<double>(max_p[0] - min_p[0]) * inverse_leaf_size[0]) + 1;
const double dy =
    std::floor(static_cast<double>(max_p[1] - min_p[1]) * inverse_leaf_size[1]) + 1;
const double dz =
    std::floor(static_cast<double>(max_p[2] - min_p[2]) * inverse_leaf_size[2]) + 1;

// Check hashing range
if (dx > max || dy > max || dz > max || dy > max / dx || dx * dy > max / dz)
  return 0;
else
  return static_cast<std::size_t>(dx * dy * dz);
```

Reply:
> Now everything are casting to `double` before multiplications, and then check in `double` domain. It won't overflow anymore in any operations

You've also lost precision here.

Reply:
🤯 I didn't know [this issue](https://stackoverflow.com/questions/52491226/converting-really-large-int-to-double-loss-of-precision-on-some-computer) before

I added additional tests in between following the logic before